# 01. Importing Libraries

In [1]:
#Import libraries
import pandas as pd
import numpy as np
import os

# 02. Data Wrangling Follow Along

In [2]:
#create folder path 
path = r'C:\Users\Victoria\Desktop\Data Analytics Career Foundry\4.0 Python Fundamentals for Data Analysts\Instacart Basket Analysis'

In [3]:
#import 'orders.csv' to create df_orders dataframe
df_orders = pd.read_csv(os.path.join(path, '02 Data', 'Original Data','orders.csv'), index_col = False)

In [4]:
#import 'products.csv' to create df_prods dataframe
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [5]:
#Permanently removing eval_et columns from df_orders
df_orders = df_orders.drop(columns = ['eval_set'])

In [6]:
#Checking 'days_since_prior_order' column for missing values
df_orders['days_since_prior_order'].value_counts(dropna = False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

In [7]:
#Changing the name of the 'order_dow' column to 'orders_day_of_the_week'
df_orders.rename(columns = {'order_dow' : 'orders_day_of_the_week'}, inplace = True)

In [8]:
#Confirming column name change
df_orders.head()

order_id  user_id  order_number  orders_day_of_the_week  order_hour_of_day  \
0   2539329        1             1                       2                  8   
1   2398795        1             2                       3                  7   
2    473747        1             3                       3                 12   
3   2254736        1             4                       4                  7   
4    431534        1             5                       4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [9]:
#Changing the data type of 'order_id' to a string 
df_orders['order_id'] = df_orders['order_id'].astype('str')

In [10]:
#Confirming data type change
df_orders['order_id'].dtype

dtype('O')

In [11]:
#Importing the data set 'departments.csv'
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

In [12]:
#Transposing df_dep dataframe into new dataframe df_dep_t
df_dep_t = df_dep.T

In [13]:
#Adding an index to df_dep_t
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [14]:
#creating an object that contains the first row of the dataframe
new_header = df_dep_t.iloc[0]

In [15]:
#Take the data under the header tow for a new df
df_dep_t_new = df_dep_t[1:]

In [16]:
#set the header row as the df header
df_dep_t_new.columns = new_header

In [17]:
#transform df_dep_t_new into a data dictionary
data_dict = df_dep_t_new.to_dict('index')

In [18]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [19]:
#Create df_snacks dataframe which shows only data from the snack department
df_snacks = df_prods[df_prods['department_id']==19]

# 03 Task

In [20]:
#Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.
#I've selected 'user_id' and will be changes it's data type to a string
df_orders['user_id'] = df_orders['user_id'].astype('str')

In [21]:
#Confirming that 'user_id' is now a string
df_orders['user_id'].dtype

dtype('O')

In [22]:
#Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the data frame.
#I've selected 'days_since_prior_order'
df_orders.rename(columns = {'days_since_prior_order' : 'days_since_last_order'}, inplace = False)

order_id user_id  order_number  orders_day_of_the_week  \
0        2539329       1             1                       2   
1        2398795       1             2                       3   
2         473747       1             3                       3   
3        2254736       1             4                       4   
4         431534       1             5                       4   
...          ...     ...           ...                     ...   
3421078  2266710  206209            10                       5   
3421079  1854736  206209            11                       4   
3421080   626363  206209            12                       1   
3421081  2977660  206209            13                       1   
3421082   272231  206209            14                       6   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
1                        7                   15.0  
2                       12                   21.0  
3                        7                   29.0  
4                       15                   28.0  
...                    ...                    ...  
3421078                 18                   29.0  
3421079                 10                   30.0  
3421080                 12                   18.0  
3421081                 12                    7.0  
3421082                 14                   30.0  

[3421083 rows x 6 columns]

In [23]:
#Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.
df_orders['order_hour_of_day'].value_counts()

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

In [24]:
#Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.
print(data_dict.get('4'))

{'department': 'produce'}


In [25]:
#Create a subset containing only the required information on breafast item sales.
#Breakfast items are department_id 14
df_breakfast = df_prods[df_prods['department_id']==14]

In [26]:
df_breakfast

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

In [27]:
#Find information on dinner party products. Find all observations from the following departments: alcohol, deli, beverages, and meat/seafood.
#I need to know the department_ids, so I will look at my data dictionary
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [28]:
#Create df_dinner_party subset using department_ids 5, 7, 12, 20
df_dinner_party = df_prods.loc[df_prods['department_id'].isin([5,7,12,20])]

In [29]:
df_dinner_party

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

In [30]:
#How many rows does df_dinner_party have?
df_dinner_party.shape

(7650, 5)

In [31]:
#Extract all information possible on customer with 'user_id' of '1'
df_userid_1 = df_orders.loc[df_orders['user_id']=='1']

In [32]:
df_userid_1

order_id user_id  order_number  orders_day_of_the_week  order_hour_of_day  \
0   2539329       1             1                       2                  8   
1   2398795       1             2                       3                  7   
2    473747       1             3                       3                 12   
3   2254736       1             4                       4                  7   
4    431534       1             5                       4                 15   
5   3367565       1             6                       2                  7   
6    550135       1             7                       1                  9   
7   3108588       1             8                       1                 14   
8   2295261       1             9                       1                 16   
9   2550362       1            10                       4                  8   
10  1187899       1            11                       4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

In [34]:
df_userid_1.describe()

order_number  orders_day_of_the_week  order_hour_of_day  \
count     11.000000               11.000000          11.000000   
mean       6.000000                2.636364          10.090909   
std        3.316625                1.286291           3.477198   
min        1.000000                1.000000           7.000000   
25%        3.500000                1.500000           7.500000   
50%        6.000000                3.000000           8.000000   
75%        8.500000                4.000000          13.000000   
max       11.000000                4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

# 04. Task Answers
2. Changed data type of 'user_id' to string
3. Renamed 'days_since_prior_order' to 'days_since-last-order'
4. Busiest hour for placing orders is 10am
5. value '4' represents 'produce
6. Create a subset for breakfast items. See df_breakfast at Out 26
7. Create a subset for dinner parties. See df_dinner_party at Out 29
8. df_dinner_parties has: 7650 rows
9. Extract all information possible on customer with 'user_id' of '1'. See Out 32
10. Basic stats of user_id 1 customr behavior. See Out 34

In [35]:
df_orders.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

In [36]:
df_dep_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))